CRUD: tasks

JknYang opened the pull request:
* Added Style and Script (CSS and Javascript) to the HTML file (including fontawesome icons)
* Added a task container for the tasks in the html
* Turned task add/edit to be hidden upon opening - CSS
* Background change when new task is opened so that it doesn't match the add/edit container - CSS
* Implemented javascript so that add/edit container opens upon click to new task
* Implemented javascript that creates the tasks
* Implemented LocalStorage for the entries/tasks, and made it a checklist [also in LocalStorage]
* Adds a date util file for date functions

p.s. comments done with AI

Closes #27, closes #30

JknYang committed:
added css link and javascript link

JknYang committed:
hid task edit/add and add task contain..

JknYang committed:
added the feature for adding/deleting/etc. tasks

kaileywong commented on source/public/scripts/main.journal.js:
```diff
@@ -0,0 +1,127 @@
+document.addEventListener("DOMContentLoaded", function () {
+    // Select the modal and various buttons
+    var modal = document.getElementById("task-modal");
```
`var` is kind of an outdated JavaScript variable declaration, since these elements aren't being reassigned it would be better to use `const` instead

kaileywong replied:
Also if you could use a bit more descriptive names for the variables like `taskModal` instead of just `modal` that would be helpful for code readbility!

kaileywong commented on source/public/scripts/main.journal.js:
```diff
@@ -0,0 +1,127 @@
+          name: taskDesc,
+          type_tag: taskType,
+          project_tag: taskProject,
+          completed: false // New tasks are not completed by default
```
Good catch that we need to track whether tasks have been completed haha - could you please update the schema + example in the `data` folder? Thanks!

kaileywong commented on source/public/scripts/main.journal.js:
```diff
@@ -0,0 +1,127 @@
+
+    // Hide the modal if the user clicks outside of it
+    window.onclick = function (event) {
+      if (event.target == modal) {
```
I think the `===` equality operator should be used here

kaileywong commented on source/public/scripts/main.journal.js:
```diff
@@ -0,0 +1,127 @@
+    // Save the task when the "Save" button is clicked
+    saveBtn.onclick = function () {
+      // Get task details from modal input fields
+      var taskDesc = document.getElementById("task-desc").value;
```
Same comment as before on using `const` instead of `var` declarations!

kaileywong commented on source/public/scripts/main.journal.js:
```diff
@@ -0,0 +1,127 @@
+      var taskProject = document.getElementById("task-project").value;
+
+      // Retrieve existing tasks from local storage or initialize as empty array
+      var tasks = JSON.parse(localStorage.getItem("tasks")) || [];
```
This works right now since we don't have infrastructure in place for entries for multiple days but would change in the future where tasks are inside an entry object (according to the schema we came up with). If you could update this to reflect that schema I think it would be helpful as we continue working on the functionalities

kaileywong commented on source/public/scripts/main.journal.js:
```diff
@@ -0,0 +1,127 @@
+      }
+
+      // Save the updated tasks array to local storage
+      localStorage.setItem("tasks", JSON.stringify(tasks));
```
This would also need to be changed to reflect the entry-task schema

kaileywong commented on source/public/scripts/main.journal.js:
```diff
@@ -0,0 +1,127 @@
+
+    // Delete the current task when the "Delete" button in the modal is clicked
+    deleteBtnModal.onclick = function () {
+      var tasks = JSON.parse(localStorage.getItem("tasks")) || [];
```
Same comment about updating according to entry-task schema

kaileywong commented on source/public/scripts/main.journal.js:
```diff
@@ -0,0 +1,127 @@
+        taskElement.className = "task-item";
+        taskElement.innerHTML = `
+          <input type="checkbox" id="task-${index}" ${task.completed ? 'checked' : ''}>
+          <label for="task-${index}">Task ${index + 1}: ${task.name}, Type: ${task.type_tag}, Project: ${task.project_tag}</label>
```
I don't think it's necessary for us to display the task number. Also it would be helpful to put the type and project tags in separate `div`s since we're planning to display them more like small badges - see the wireframe:
 
![journal_page](https://github.com/cse110-sp24-group33/cse110-sp24-group33/assets/61530252/bbeae804-6be0-4233-84b6-bfdec9792d05)

kaileywong replied:
Also then maybe instead of just setting the `innerHTML` it would be better for readability to create new elements and append each into the `div`

kaileywong commented on source/public/scripts/main.journal.js:
```diff
@@ -0,0 +1,127 @@
+        var checkbox = taskElement.querySelector(`input[type="checkbox"]`);
+        checkbox.addEventListener('change', function () {
+          tasks[index].completed = this.checked; // Update completed status
+          localStorage.setItem("tasks", JSON.stringify(tasks)); // Save updated tasks array to local storage
```
Since this line is being called a lot, maybe for readability turning it into a small function like `updateTasks()` would be helpful?

kaileywong reviewed (changes requested):
Really good work on the tasks functionality! Just left some comments for small considerations with things like vars, schemas, and displays. Also, the issue you're working on here includes sentiment functionality - would you prefer continuing to work on that on the same branch here or merging this first and creating another PR for that later?

JknYang commented:
Hi Kailey, I don't have a preference to which branch I work on for further issues. I'll work on whatever that is least inconvenient for the rest of the team.

kaileywong commented:
Ok, I think it would result in less conflicts as others are working to finish merging up this portion for the tasks and start a new PR for sentiment

JknYang committed:
Var --> const

JknYang committed:
Added "completed" to the schema

JknYang committed:
Added some changes

Found the changes minus "This would also need to be changed to reflect the entry-task schema" and "updateTask()"

kaileywong committed:
update schemas, cleanup tasks functionality, separate date util, add fa icons

kaileywong committed:
fix next day btn

kaileywong committed:
add home page link
